tests/controllers/test_elasticsearch_controller.py

```diff
@@ -0,0 +1,156 @@
+    with patch("morpheus.controllers.elasticsearch_controller.Elasticsearch", Mock(spec=Elasticsearch)):
+        controller = ElasticsearchController("test_index", connection_kwargs, refresh_period_secs=10)
+        yield controller
+
+
+@pytest.fixture(scope="module")
+def connection_kwargs():
+    kwargs = {"hosts": [{"host": "localhost", "port": 9200, "scheme": "http"}]}
+    yield kwargs
+
+
+@pytest.mark.use_python
+def test_constructor(mock_es_controller, connection_kwargs):
+    assert mock_es_controller._index == "test_index"
+    assert mock_es_controller._connection_kwargs == connection_kwargs
+    assert mock_es_controller._raise_on_exception is False
+    assert mock_es_controller._refresh_period_secs == 10
+    assert mock_es_controller._client is not None
+
+
+@pytest.mark.use_python
+def test_apply_derive_params_func(mock_es_controller):
+    pickled_func_str = str(pickle.dumps(mock_derive_params), encoding="latin1")
+
+    # Create pickled_func_config
+    pickled_func_config = {
+        "pickled_func_str": pickled_func_str,  # Pickled function string
+        "encoding": "latin1"
+    }
+
+    # Apply the mock function and check if connection_kwargs is updated
+    mock_es_controller._apply_derive_params_func(pickled_func_config)
+    assert mock_es_controller._connection_kwargs == {
+        "hosts": [{
+            "host": "localhost", "port": 9200, "scheme": "http"
+        }], "retry_on_status": 3, "retry_on_timeout": 30
+    }
+
+
+@pytest.mark.use_python
+def test_refresh_client_force(mock_es_controller):
+    # Simulate a force refresh
+    mock_es_controller.refresh_client(force=True)
+
+    assert mock_es_controller._client is not None
+    assert mock_es_controller._last_refresh_time > 0
+    assert isinstance(mock_es_controller._client, Mock)
+
+
+@pytest.mark.use_python
+def test_refresh_client_not_needed(mock_es_controller):
+    # Set last refresh time to a recent time
+    current_time = time.time()
+    mock_es_controller._last_refresh_time = current_time
+
+    # Simulate a refresh not needed scenario
+    mock_es_controller.refresh_client()
+    # Assert client is not None
+    assert mock_es_controller._client is not None
+    # Assert last_refresh_time is unchanged
+    assert mock_es_controller._last_refresh_time == current_time
+    # Assert client type remains the same
+    assert isinstance(mock_es_controller._client, Mock)
+
+
+@pytest.mark.use_python
+def test_refresh_client_needed(mock_es_controller):
+    # Set last refresh time to a recent time
+    current_time = time.time()
+    mock_es_controller._refresh_period_secs = 1
+    mock_es_controller._last_refresh_time = current_time
+    time.sleep(1)
+
+    # Simulate a refresh needed scenario
+    mock_es_controller.refresh_client()
+
+    # Assert client is not None
+    assert mock_es_controller._client is not None
+    # Assert last_refresh_time is changed
+    assert mock_es_controller._last_refresh_time > current_time
```

Comment:
since we are patching the client, `orig_client` and the refreshed `client` end up referring to the same instance. I'm asserting the return value of the `refresh_client`

Reply:
Ah true. Then use a return value on `refresh_client()` instead.